File: src/my_cpp_pkg/src/vision_utils.cpp

#include "my_cpp_pkg/vision_utils.hpp"
#include <opencv2/calib3d.hpp>

namespace vision_utils {

Matrix_t ComputeHomography(const std::vector<cv::Point2d>& cv_reference_points,
                           const std::vector<cv::Point2d>& cv_current_points,
                           const Matrix_t& K) {
    std::vector<cv::Point2f> ref_pts, curr_pts;
    for (const auto& pt : cv_reference_points) {
        ref_pts.emplace_back(static_cast<float>(pt.x), static_cast<float>(pt.y));
    }
    for (const auto& pt : cv_current_points) {
        curr_pts.emplace_back(static_cast<float>(pt.x), static_cast<float>(pt.y));
    }

    cv::Mat H_cv = cv::findHomography(ref_pts, curr_pts, cv::RANSAC);
    Matrix_t H_eigen;
    cv::cv2eigen(H_cv, H_eigen);
    return K.inverse() * H_eigen * K;
}

void RecoverFromHomography(const Matrix_t& homography,
                           Matrix_t& R,
                           Vector3D_t& t,
                           Vector3D_t& n,
                           FPTYPE& distancePlane,
                           int current_iteration,
                           homographySolution& homography_solution) {
    Eigen::JacobiSVD<Matrix_t> svd(homography, Eigen::ComputeFullU | Eigen::ComputeFullV);
    Eigen::Vector3d singular_values = svd.singularValues();

    const Matrix_t& U = svd.matrixU();
    const Matrix_t& V = svd.matrixV();

    Matrix_t W;
    W << 1, 0, 0,
         0, 1, 0,
         0, 0, singular_values[0]/singular_values[1];

    Matrix_t R1 = U * W * V.transpose();
    Matrix_t R2 = U * W.transpose() * V.transpose();

    std::vector<Matrix_t> possible_rotations = { R1, R2 };
    std::vector<Vector3D_t> possible_translations, possible_normals;

    for (const auto& R_candidate : possible_rotations) {
        Vector3D_t t_candidate = U.col(2) * (singular_values[0] + singular_values[1]) / 2.0;
        Vector3D_t n_candidate = V.col(2).normalized();

        possible_translations.push_back(t_candidate);
        possible_normals.push_back(n_candidate);
    }

    // Simplified selection: choose first solution for demonstration
    R = possible_rotations[0];
    t = possible_translations[0];
    n = possible_normals[0];
    distancePlane = 1.0; // Placeholder
    homography_solution = homographySolution::SOLUTION_1;
}

void EstimateHomography(const cv::Mat& refImg,
                        const cv::Mat& currImg,
                        const Matrix_t& K,
                        Matrix_t& H,
                        int counter) {
    cv::Ptr<cv::ORB> orb = cv::ORB::create(500);
    std::vector<cv::KeyPoint> kp1, kp2;
    cv::Mat des1, des2;
    orb->detectAndCompute(refImg, cv::noArray(), kp1, des1);
    orb->detectAndCompute(currImg, cv::noArray(), kp2, des2);

    cv::BFMatcher matcher(cv::NORM_HAMMING);
    std::vector<cv::DMatch> matches;
    matcher.match(des1, des2, matches);

    std::vector<cv::Point2d> ref_pts, curr_pts;
    for (const auto& m : matches) {
        ref_pts.push_back(kp1[m.queryIdx].pt);
        curr_pts.push_back(kp2[m.trainIdx].pt);
    }

    H = ComputeHomography(ref_pts, curr_pts, K);
}

void Rodriguez(const Matrix_t& R, Vector3D_t& u) {
    cv::Mat R_cv;
    cv::eigen2cv(R, R_cv);
    cv::Mat rvec;
    cv::Rodrigues(R_cv, rvec);
    cv::cv2eigen(rvec, u);
}

void PBVSController(const Matrix_t& R,
                    const Vector3D_t& t,
                    const Vector3D_t& u,
                    Vector3D_t& Uv,
                    Vector3D_t& Uw,
                    const FPTYPE lambdav_,
                    const FPTYPE lambdaw_) {
    Uv = -lambdav_ * R.transpose() * t;
    Uw = -lambdaw_ * u;
}

Point2D_t NormalizePoint(const Point2D_t& point, const Matrix_t& Kinv_) {
    Vector3D_t homog_point(point.x(), point.y(), 1.0);
    Vector3D_t normalized = Kinv_ * homog_point;
    return Point2D_t(normalized.x() / normalized.z(), normalized.y() / normalized.z());
}

} // namespace vision_utils

File: src/my_py_pkg/my_py_pkg/vision_utils.py
import numpy as np
import cv2
from numpy.linalg import inv, svd

def unitize(x, y):
    """Normalize 2D vector"""
    magnitude = np.sqrt(x**2 + y**2)
    return x/magnitude, y/magnitude

def vex(A):
    """Convert skew-symmetric matrix to vector"""
    if A.shape != (3, 3):
        raise ValueError("vex: expects 3x3 matrices as input")
    return np.array([A[2,1]-A[1,2],
                     A[0,2]-A[2,0],
                     A[1,0]-A[0,1]]) * 0.5

def compute_homography(ref_pts, curr_pts, K):
    """Compute normalized homography"""
    H, _ = cv2.findHomography(ref_pts, curr_pts, cv2.RANSAC)
    return inv(K) @ H @ K  # Use matrix operator @ instead of dot()

def recover_from_homography(H, K, iter_count, prev_solution):
    """Recover pose from homography matrix"""
    U, S, V_T = svd(H) # Use svd directly from numpy.linalg
    V = V_T.T
    s = S[0] + S[1]
    t1 = np.array([U[:, 2]]).T * s / 2.0
    t2 = -np.array([U[:, 2]]).T * s / 2.0
    W = np.diag([1, 1, (S[0] / S[1])])
    W_inv = W.T
    R1 = U @ W @ V_T
    R2 = U @ W_inv @ V_T
    n1 = np.array([V[:, 2]]).T
    n2 = - np.array([V[:, 2]]).T # Normal vector n2 = -n1

    d = 1 # Since Zref = Zcurrent

    # Ensure rotation matrices are proper rotations (determinant +1) and normals point towards camera.
    if np.linalg.det(R1) < 0:
        R1 = -R1
    if np.linalg.det(R2) < 0:
        R2 = -R2
    if n1[2] < 0:
        n1 = -n1
    if n2[2] < 0:
        n2 = -n2

    possible_solutions = [{'R': R1, 't': t1, 'n': n1, 'solution_tag': 1},
                           {'R': R2, 't': t1, 'n': n2, 'solution_tag': 2},
                           {'R': R1, 't': t2, 'n': n2, 'solution_tag': 3},
                           {'R': R2, 't': t2, 'n': n1, 'solution_tag': 4}]

    # In a real scenario, you would have to implement a more robust solution selection
    # based on reprojection error and/or cheirality checks.
    # Here, for simplicity, we just return the first valid solution.
    # A basic heuristic could be to check if the translation is not NaN.
    # And also check if the recovered rotation is close to a rotation matrix.

    best_solution = None
    min_error = float('inf')

    for sol in possible_solutions:
        R_candidate = sol['R']
        t_candidate = sol['t']
        n_candidate = sol['n']

        if not np.isnan(np.sum(t_candidate)): # Basic NaN check for translation
            if best_solution is None:
                 best_solution = sol # Select the first valid solution as best for now.


    if best_solution is not None:
        R = best_solution['R']
        t = best_solution['t']
        n = best_solution['n']
        solution_tag = best_solution['solution_tag']
    else:
        R = np.eye(3)
        t = np.zeros((3, 1))
        n = np.zeros(3)
        d = 0.0
        solution_tag = 0 # Indicate no valid solution found

    return R, t, n, d, solution_tag


def estimate_homography(ref_img, curr_img, K, counter):
    """Feature-based homography estimation"""
    orb = cv2.ORB_create(600)
    kp1, des1 = orb.detectAndCompute(ref_img, None)
    kp2, des2 = orb.detectAndCompute(curr_img, None)

    bf = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=True)
    matches = sorted(bf.match(des1, des2), key=lambda x: x.distance)[:50]

    # Convert keypoints to numpy arrays
    ref_pts = np.float32([kp1[m.queryIdx].pt for m in matches])
    curr_pts = np.float32([kp2[m.trainIdx].pt for m in matches])

    if len(ref_pts) < 4 or len(curr_pts) < 4:
        return None

    H, _ = cv2.findHomography(ref_pts, curr_pts, cv2.RANSAC, 5.0)
    return (inv(K) @ H @ K) / H[1,1]  # Modern matrix operator

def rodriguez(R):
    u, _ = cv2.Rodrigues(R)
    return u

def pbvs_controller(R, t, u, lambdav, lambdaw):
    Uv = -lambdav * np.dot(R.T, t)
    Uw = -lambdaw * u
    return Uv, Uw

def normalize_point(point, Kinv):
    norm_point = np.dot(Kinv, np.hstack((point, [1])))
    return norm_point[:2] / norm_point[2]
